lib/semantics/check-omp-structure.h

```diff
@@ -72,39 +108,41 @@ class OmpStructureChecker : public virtual BaseChecker {
   void Enter(const parser::OmpReductionClause &);
   void Enter(const parser::OmpScheduleClause &);
 
-  void EmptyAllowed() { allowedClauses_ = {}; }
-  void SetAllowed(std::set<std::type_index> allowedClauses) {
-    allowedClauses_ = allowedClauses;
+  bool HasInvalidCloselyNestedRegion(
+      const parser::CharBlock &, const OmpDirectiveEnumType &);
+  void CheckAllowed(const OmpClauseEnum &);
+
+private:
+  struct OmpContext {
+    parser::CharBlock currentDirectiveSource{nullptr};
+    parser::CharBlock currentClauseSource{nullptr};
+    OmpDirectiveEnum currentDirectiveEnum;
+    OmpClauseEnumType allowedClauses;
+    OmpClauseEnumType allowedOnceClauses;
+    OmpClauseEnumType seenClauses;
+  };
+  const OmpContext &GetOmpContext() const { return ompContext_.back(); }
+  void SetOmpContextDirectiveSource(const parser::CharBlock &directive) {
+    ompContext_.back().currentDirectiveSource = directive;
   }
-  template<typename N> void CheckAllowed(const N &node) {
-    if (allowedClauses_.empty()) {
-      context_.Say(
-          currentClauseSource_, "Internal: allowed clauses not set"_err_en_US);
-    }
-    auto it{allowedClauses_.find(typeid(node))};
-    if (it == allowedClauses_.end()) {
-      context_.Say(currentClauseSource_, "'%s' not allowed in OMP %s"_err_en_US,
-          parser::ToUpperCaseLetters(currentClauseSource_.ToString()),
-          parser::ToUpperCaseLetters(currentDirectiveSource_.ToString()));
-    }
+  void SetOmpContextClauseSource(const parser::CharBlock &clause) {
+    ompContext_.back().currentClauseSource = clause;
```

Comment:
Rather than defining all of these setters, why not change `GetOmpContext()` to return a non-const ref. Then replace `SetOmpContextClauseSource(x);` with `GetOmpContext().currentClauseSource = x;`.

Reply:
The suite of `Set` routines define a limited API for mutation and might be safer, though.

Reply:
I think there is a setter for each data member.

Reply:
Sure, but that's all that they can do; a reference could be abused in other ways.